lalrpop-test/src/main.rs

```diff
@@ -281,6 +287,65 @@ fn error_test1() {
 }
 
 #[test]
+fn error_recovery_eof() {
+    let errors = RefCell::new(vec![]);
+    util::test(|v| error_recovery::parse_Item(&errors, v), "--", '!'.to_string());
+
+    assert_eq!(errors.borrow().len(), 1);
+    assert_eq!(errors.borrow()[0], ParseError::UnrecognizedToken {
+        token: None,
+        expected: vec![],
+    });
+}
+
+#[test]
+fn error_recovery_eof_without_recovery() {
+    let errors = RefCell::new(vec![]);
+    let tokens = util::tok::tokenize("-").into_iter().map(|t| t.1);
+    let result = error_recovery::parse_Item(&errors, tokens);
+    assert_eq!(result, Err(ParseError::UnrecognizedToken {
+        token: None,
+        expected: vec![],
+    }));
+}
+
+#[test]
+fn error_recovery_extra_token() {
+    let errors = RefCell::new(vec![]);
+    util::test(|v| error_recovery::parse_Item(&errors, v), "(++)", "()".to_string());
+
+    assert_eq!(errors.borrow().len(), 1);
+    assert_eq!(errors.borrow()[0], ParseError::UnrecognizedToken {
+        token: Some(((), Tok::Plus,())),
+        expected: vec![],
+    });
+}
+
+#[test]
+fn error_recovery_dont_drop_unrecognized_token() {
+    let errors = RefCell::new(vec![]);
+    util::test(|v| error_recovery::parse_Item(&errors, v), "(--)", "!".to_string());
```

Comment:
Hmm, I sort of expected this to yield `(!)` as the result -- i.e., we start with the `(`, then push an item state, then encounter an error on the `)`, so we push `error`, but I'd expet the `)` to be in the follow-set, so we'd produce a `!` that gets wrapped in `()`, no?

In general, it would be good to add some comments to each test demonstrating the flow. The names are great but comments are good too.